src/System.Data.SqlClient/src/System/Data/SqlClient/SqlFileStream.Windows.cs

```diff
@@ -0,0 +1,979 @@
+
+        // path length limitations:
+        // 1. path length storage (in bytes) in UNICODE_STRING is limited to UInt16.MaxValue bytes = Int16.MaxValue chars
+        // 2. GetFullPathName API of kernel32 does not accept paths with length (in chars) greater than 32766
+        //  (32766 is actually Int16.MaxValue - 1, while (-1) is for NULL termination)
+        // We must check for the lowest value between the the two
+        private const int MaxWin32PathLength = short.MaxValue - 1;
+
+        [Conditional("DEBUG")]
+        static private void AssertPathFormat(string path)
+        {
+            Debug.Assert(path != null);
+            Debug.Assert(path == path.Trim());
+            Debug.Assert(path.Length > 0);
+            Debug.Assert(path.Length <= MaxWin32PathLength);
+            Debug.Assert(path.IndexOfAny(s_invalidPathChars) < 0);
+            Debug.Assert(path.StartsWith(@"\\", StringComparison.OrdinalIgnoreCase));
+            Debug.Assert(!path.StartsWith(@"\\.\", StringComparison.Ordinal));
+        }
+
+        // SQLBUVSTS01 bugs 192677 and 193221: we cannot use System.IO.Path.GetFullPath for two reasons:
+        // * it requires PathDiscovery permissions, which is unnecessary for SqlFileStream since we are dealing with network path
+        // * it is limited to 260 length while in our case file path can be much longer
+        // To overcome the above limitations we decided to use GetFullPathName function from kernel32.dll
+        [ResourceExposure(ResourceScope.Machine)]
+        [ResourceConsumption(ResourceScope.Machine)]
+        static private string GetFullPathInternal(string path)
+        {
+            //-----------------------------------------------------------------
+            // precondition validation should be validated by callers of this method
+            // NOTE: if this method moves elsewhere, this assert should become an actual runtime check
+            // as the implicit assumptions here cannot be relied upon in an inter-class context
+            Debug.Assert(path != null);
+
+            // remove leading and trailing whitespace
+            path = path.Trim();
+            if (path.Length == 0)
+            {
+                throw ADP.Argument(SR.GetString(SR.SqlFileStream_InvalidPath), "path");
+            }
+
+            // check for the path length before we normalize it with GetFullPathName
+            if (path.Length > MaxWin32PathLength)
+            {
+                // cannot use PathTooLongException here since our length limit is 32K while
+                // PathTooLongException error message states that the path should be limited to 260
+                throw ADP.Argument(SR.GetString(SR.SqlFileStream_InvalidPath), "path");
+            }
+
+            // GetFullPathName does not check for invalid characters so we still have to validate them before
+            if (path.IndexOfAny(s_invalidPathChars) >= 0)
```

Comment:
I'd recommend not checking for invalid chars, we don't any more. I can live with this staying, however.